cmd/test_gen.go

```diff
@@ -0,0 +1,314 @@
+	pb := utils.NewProgressBar("🖨 Generating tests for models in graph", graph.Len())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	var testSugs TestSuggestions
+	testSugs.Init()
+
+	err := graph.Execute(func(file *fs.File) error {
+		if file.Type == fs.ModelFile {
+			testSuggestions, err := generateTestsForModel(ctx, file)
+			if err != nil {
+				pb.Stop()
+				if err != context.Canceled {
+					fmt.Printf("❌ %s\n", err)
+				}
+				cancel()
+				return err
+			}
+			testSugs.SetSuggestion(file.Name, testSuggestions)
+		}
+
+		pb.Increment()
+		return nil
+	}, config.NumberThreads(), pb)
+
+	if err != nil {
+		return err
+	}
+	pb.Stop()
+
+	err = userPromptTests(graph, testSugs.suggestions)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// generateTestsForModel generates tests for model and writes yml schema for modelName.
+func generateTestsForModel(ctx context.Context, file *fs.File) (map[string][]string, error) {
+	target, err := file.GetTarget()
+	if err != nil {
+		fmt.Println("could not get target for schema")
+		return nil, err
+	}
+	fmt.Println("\n🎯 Target for retrieving schema:", target.ProjectID+"."+target.DataSet)
+
+	// retrieve columns from BigQuery
+	bqColumns, err := getColumnsForModel(ctx, file.Name, target)
+	if err != nil {
+		fmt.Println("Could not retrieve schema")
+		return nil, err
+	}
+	fmt.Println("✅ BQ Schema retrieved. Number of columns in BQ table:", len(bqColumns))
+
+	// iterate through functions which return test sql and definition
+	testFuncs := []func(string, string, string, string) (string, string){
+		schemaTestMacros.Test_not_null_macro,
+		schemaTestMacros.Test_unique_macro,
+	}
+
+	var allTestQueries []ColumnTestQuery
+	for _, col := range bqColumns {
+		for _, test := range testFuncs {
+			testQuery, testName := test(target.ProjectID, target.DataSet, file.Name, col)
+			allTestQueries = append(allTestQueries, ColumnTestQuery{
+				Column:    col,
+				TestName:  testName,
+				TestQuery: testQuery,
+			})
+		}
+	}
+
+	passedTestQueries, err := runQueriesParallel(ctx, target, allTestQueries)
+	if err != nil {
+		return nil, err
+	}
+	updateSchemaFile(passedTestQueries, file)
+
+	return passedTestQueries, nil
+}
+
+func runQueriesParallel(ctx context.Context, target *config.Target, allTestQueries []ColumnTestQuery) (map[string][]string, error) {
+	// number of parallel query runners
+	numQueryRunners := 100
+
+	queries := make(chan ColumnTestQuery)
+	go func() {
+		for _, q := range allTestQueries {
+			queries <- q
+		}
+		close(queries)
+	}()
+
+	out := make(chan ColumnTestQuery, len(allTestQueries))
+	errs := make(chan error, len(allTestQueries))
+	wg := sync.WaitGroup{}
+
+	for i := 0; i < numQueryRunners; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			for query := range queries {
+				evaluateTestQuery(ctx, target, query, out, errs, i)
+			}
+		}(i)
+	}
+
+	go func() {
+		wg.Wait()
+		close(out)
+		close(errs)
+	}()
+
+	if len(errs) > 0 {
```

Comment:
I _think_ this might not work as you expect:
Everything above this are in goroutines, so when it reaches here, all the goroutines might still be running (and may have errors later, but len(errs) is 0).

What we want to do here is probably wait for the `errs` channel to closed on line 211 -- it's guaranteed that the waitgroup is complete and all the errors that can happened has happened.

Perhaps we can read the first error and fail early
```
// (2) read first error only
if err := <-errs; err != nil {
  // there is at least one error, but we ignore the rest
  return nil, err
}
```
Or alternatively, add a `for err := range errs { ... }` to print out all errors

Reply:
Cool this is useful - I think the first error is actually fine here. So will add this